File: DesignQueueWithStack.cs
//used two stack one to push the elements and other stack to pop the elements
//Whlie pushing time complexity is O(1) as element is directly push to stack
//while pop time complexity is O(1) as we only copy to other stack when that stack is empty
//so average case time complexity is O(1)

//Total TC: O(1)

public class MyQueue {
    private Stack<int> mainStack;
    private Stack<int> popStack;
    public MyQueue() {
        mainStack= new Stack<int>();
        popStack= new Stack<int>();
    }
    
    //TC: O(1)
    public void Push(int x) {
        mainStack.Push(x);
    }
    
    //TC: O(1)
    public int Pop() {
        Peek();  
        return  popStack.Pop();      
    }
    
    //TC: O(1)
    public int Peek() {
       if(popStack.Count == 0){
            while(mainStack.Count!=0){
                popStack.Push(mainStack.Pop());
            }
        }
        return popStack.Peek();
    }
    
    //TC: O(1)
    public bool Empty() {
        if(mainStack.Count == 0 && popStack.Count  == 0){
            return true;
        }
        return false;
    }
}

File: designQueueWithStack.java
// Time complexity: O(1)
// Space Complexity: O(1)


import java.util.Stack;

class MyQueue {
        private Stack<Integer> in;
        private Stack<Integer> out;
    public MyQueue() {
        in = new Stack<>();
        out = new Stack<>();
    }
    
    public void push(int x) {
        in.push(x);
    }
    
    public int pop() {
        if(out.isEmpty()){
            while(!in.isEmpty()){
                out.push(in.pop());
            }
        }
        return out.pop();
    }
    
    public int peek() {
         if(out.isEmpty()){
            while(!in.isEmpty()){
                out.push(in.pop());
            }
        }
        return out.peek();
    }
    
    public boolean empty() {
        return (in.isEmpty() && out.isEmpty());
    }

    public static void main(String args[]){
           MyQueue obj = new MyQueue();
           obj.push(2);
           obj.push(3);
           obj.push(4);
           int param_2 = obj.pop();
           System.out.println(param_2);
           int param_3 = obj.peek();
           System.out.println(param_3);
           boolean param_4 = obj.empty();
           System.out.println(param_4);
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */